sql/src/main/java/io/crate/metadata/Functions.java

```diff
@@ -101,4 +161,39 @@ public TableFunctionImplementation getTableFunctionSafe(String name) {
         }
         return tableFunctionImplementation;
     }
+
+    private static class GeneratedFunctionResolver implements FunctionResolver {
+
+        private final String name;
+        private final LinkedHashMap<FunctionIdent, FunctionImplementation> functions = new LinkedHashMap<>();
+
+        GeneratedFunctionResolver(String name) {
+            this.name = name;
+        }
+
+        public void add(FunctionIdent ident, FunctionImplementation implementation) {
+            functions.put(ident, implementation);
+        }
+
+        @Override
+        public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
+            FunctionIdent ident = new FunctionIdent(name, dataTypes);
+            return functions.get(ident);
+        }
+
+        @Override
+        public List<Signature> signatures() {
+            // TODO: sort list
+            return Lists.transform(Lists.newArrayList(functions.keySet()), new Function<FunctionIdent, Signature>() {
```

Comment:
I think this would be more efficient if done non-lazy. It's creating a copy of the keys. So there is a list-copy + transform view + function instance allocation.
I'd just create a new list and add the signatures while looping over the keyset.